### DISCUSSION TITLE
Add valve device class (either to switch or cover)

### DISCUSSION THREAD

**Discussion by cibernox** [home-assistant/architecture#971] [Entity Models] [answered]
There are many smart valves in the market. Some are actual smart valves and others are actuators with stepper motors that actually turn good old manual valves (water, gas, etc...) but in essence they are the same. Other appliances have valves inside too. My heat recuperator has also a valve to bypass the heat exchange and inject fresh air directly from the outside in summer nights (commonly referred to as "freecooling"). Swimming pools have plenty of valves.

The bottom lines is that valves are pretty common.

It would be nice to be able to represent those better than a generic switch. It could be done in a couple ways:
1. Add a new device class to switches (right the only options are switches and outlets)
2. Consider valves like a kind of device class for covers (which might be handy if valves support being 50% open, although mines don't. I never came across half-opened valves in a home setup, but possibly in industrial setups they exist). This kind of aligns well with how this would be controlled with voice commands. When we speak about switches we say "turn on/off", but for valves the natural word is "open/close".

Whatever the preferred path is, the default icon for valves could be set to `mdi:valve` and the open/closed states could use `mdi:valve-open` and `mdi:valve-closed` automatically.

It would be a nice touch, that's all.

I've added device classes before, so I have a rough understanding of what the required changes would be, I can probably do the work if you agree this is desirable and provide feedback on what the best approach would be.

**Comment by frenck** [comment]
As with anything we add, there must be a use case within the Core codebase.

Which integration would benefit from this today?

../Frenck

**Reply by elupus** [reply]
Of mine:
- Nibe heatpump (With reservation, its not clear why you would manually control)
- Gardena Bluetooth

**Reply by thecode** [reply]
For Shelly we could use a valve type for the Shelly Gas as a cover or a switch, this device has a valve for controlling a Gas Valve.
Today we used a switch class and the opening/closing states are represented as a status sensor.
I think a cover which is shown as a valve with support for open/close/opening/closing states is a better representation, although the name of the platform - "cover" doesn't fit valves while switch does it much better without support for closing & opening.

**Reply by Jc2k** [reply]
A valve in homekit is basically a switch. For us, a device class would be great. We wouldn't be able to use opening/closing states if they existed.

**Reply by cibernox** [reply]
For me that would cover all my uses. If no one can provide examples of valves that open half way, I’d feel inclined towards modeling valves as switches with a special device class. 

Since valves open quite fast, transients states likely opening and closing seem nite to have but not critical.

**Comment by thecode** [comment]
> Since valves open quite fast, transients states likely opening and closing seem nite to have but not critical.

This is a wrong assumption, I already stated earlier that I don't mind valve being a Switch, but I would not considered the speed as a factor. I have a cover that is only 60cm long and it open/closes faster than most valves I know.

While writing this comment I also realized that Switch is only used for devices that can turn on or off until now (https://www.home-assistant.io/integrations/switch/#device-class) and cover is used for devices that can open or close (https://www.home-assistant.io/integrations/cover/#device-class), so I am updating my preference that valve should be based on cover, otherwise we break the use of a switch platform even if a specific valve doesn't support opening/closing states (same as done for covers without intermediate states)

**Reply by Jc2k** [reply]
We already have valves in HA as switches, but with no device class. So changing to cover would be a much bigger breaking change for those devices.

I guess we should stand back and ask what is a valve? In homekit something like an Eve Aqua is a valve. The valve type is also used by showers. The Gardena Bluetooth that was mentioned is like an Eve Aqua.

To an engineer type of person both of these devices clearly actuate a valve.

But to most people, you don't open or close the tap and you don't open or close the shower. You turn both on and off.

You also don't cover a tap or shower with a valve.

I'd rather extend the number entity for % open (perhaps adding the concept of a target value and a current number to it?) than to use cover because it just feels weird to use cover for something that is not a cover just because it has opening and closing states.

**Comment by thecode** [comment]
> I guess we should stand back and ask what is a valve? In homekit something like an Eve Aqua is a valve. The valve type is also used by showers. The Gardena Bluetooth that was mentioned is like an Eve Aqua.
> 
> To an engineer type of person both of these devices clearly actuate a valve.
> 
> But to most people, you don't open or close the tap and you don't open or close the shower. You turn both on and off.
> 
> You also don't cover a tap or shower with a valve.
> 

Actually if looking from assistants point of view (at least what I understand with Google Assistant) you say "Open shower valve" and not "Turn on shower valve", but as said earlier this is not a clear cut, the platform `cover` is not an ideal fit mostly due to the name of it, not the behaviour. There is also no conflict IMO if the same device class is used in different platform types (switch, cover). I don't have valves today, I just stated a use case with the Shelly valve (which we also modeled as a Switch with a sensor for status). However when I did use a gate with a Switch for a gate (historically) I would need to tell google to turn on the gate instead of telling it to open the gate.

**Reply by cibernox** [reply]
I see the point of having valves that only support on/off modeled as a device class for switches and valves that have either intermediate states or transient states (opening/closing) be modeled as a device class for covers. IDK if the core team will find that palatable tho, it adds a possible source of confusion for users, specially non technical ones. 

The option of having valves being a top level integration, like lights, would be simpler from a user’s perspective (switch as valve, much like we have an option to have switches pose as lights today) but seems a bit overkill.

**Comment by BradleyFord** [comment]
ZigBee valves are shown in %, this one takes about 30sec to go from open to
closed.
But I can also set any percentage.

**Comment by cibernox** [comment]
So, to sumarize the discussions so far:
- Some people report that their valves are binary, they can either be open or closed. This seems to be the case for homekit valves. This is also the case for the valves I've seen.
- Others seem to have valves that have % of "openess". That obviously cannot be modelled as a simple switch. A cover would model this better (despite of being a terrible name for a valve. But it's not a great name for a door either, which a valve for people 😄 ).

The options we have, shall we decide to make valves nicer, are four:
1. Model all valves as a device class for switches and dispend the fact that _some_ valves can open half way, and also with transient states like opening/closing.
2. Model all valves as a device class for covers, knowing that most valves will not make use of % and transient states.
3. Allow `device_class: "valve"` for both switches and covers. It is a bit more complex to document/teach, but depending on the kind of valve you have you can choose one or the other.
4. Make valves a top level integration, at the same level as switches and covers. A sibling of them.

**Reply by Jc2k** [reply]
(In reply to 4) To avoid this happening again (we have another openy-closey thing that's not a blind or a valve), maybe something more generic than valve?

"Position" represents the state of a physical object between point A and point B.

"Transition" represents the state of a transition between 2 states.

Maybe "number" with some device classes and support for target state vs actual state is enough?

**Comment by BradleyFord** [comment]
Note: as an oil refinery guy...and this isn't all relevant to home usage.
- We normally visually distinguish the valve types on our control system
UI, so it is visually easy to see what has a "positioner" (device that can
set precise % position) and track the "travel" of the valve. Generally
speaking these are all in %, that is converted from 0-10v, but also could
be directly sent as a % from the smart positioner
- On/Off valves are generally sent as discrete outputs, in some cases a
switch at fully closed/open as an input to confirm the final state.
- On/Off and positioner based valves are treated very differently due to
different control mechanisms, different usage,

Originally when answering this question I was going to suggest
combining into one; but as I reflect more I feel it actually makes more
sense to represent them as 2 different types. My basis for this is that it
would make it much simpler to implement and use.


From Jc2k comment:
Maybe we should have 2 types of "openy-closey" thing:
1. Discrete (on/off)
2. Analog (%)
This would then work for various types of blinds, valves, garage doors, etc.

**Comment by emontnemery** [comment]
I agree we should do 4, and we should do `valve`, not `openy_closey`: `switch` and `number` already serve as unspecified "openy-closey" things; `switch` is binary and `number` has a position, I don't think we need to add one more such domain.

A `valve` entity would probably be very similar to `cover` which supports either binary state or a numeric state, but without tilt and with different device classes.

@cibernox if you agree, could you make a proposal for a `valve` entity:
- What states should a valve have
- What state attributes should a valve have
- What supported features should it have
- What services should it have
- List initial supported device classes

One detail which we should not copy from `cover` is that it requires covers which know their position to implement an `is_closed` property; this should instead be calculated by the `valve` base class to enforce the definitions of open/closed in relation to the valve position.

**Reply by cibernox** [reply]
@emontnemery sure thing. I'll start another discussion to flesh out how valves should behave.
As for device classes, this it depends a lot on how into the weeds we want to go. We could have water valves, gas valves, air valves.... or we could go more in detail and consider as valves things like showers, sprinklers, etc... 
We can flesh this out in the new discussion (tonight, at some point)

**Comment by cibernox** [comment, marked as answer]
Instead of a valve device class for switch and cover, we should do a valve integration.

Discussion about the specifics of the valve integration have been moved to https://github.com/home-assistant/architecture/discussions/975
